include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp

```diff
@@ -0,0 +1,519 @@
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Provides the algorithm to search in an index using search schemes.
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <range/v3/view/slice.hpp>
+
+#include <seqan3/core/metafunction/transformation_trait_or.hpp>
+#include <seqan3/search/algorithm/detail/search_common.hpp>
+#include <seqan3/search/algorithm/detail/search_scheme_precomputed.hpp>
+#include <seqan3/search/algorithm/detail/search_trivial.hpp>
+#include <seqan3/search/fm_index/concept.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\addtogroup search
+ * \{
+ */
+
+/*!\brief Computes a (non-optimal) search scheme. Currently the generated search scheme represents trivial
+ *        backtracking.
+ * \param[in] min_error Minimum number of errors allowed.
+ * \param[in] max_error Maximum number of errors allowed.
+ */
+inline std::vector<search_dyn> compute_ss(uint8_t const min_error, uint8_t const max_error)
+{
+    // TODO: Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
+
+/*!\brief Returns for each search the cumulative length of blocks in the order of blocks in each search and the
+ *        starting position of the first block in the query sequence.
+ * \param[in] search_scheme Search scheme that will be used for searching.
+ * \param[in] query_length Length of the query that will be searched in an index.
+ */
+template <typename search_scheme_t>
+inline auto search_scheme_block_info(search_scheme_t const & search_scheme, uint64_t const query_length)
+{
+    using blocks_length_type = typename search_scheme_t::value_type::blocks_length_type;
+
+    // Either store information in an array (for search schemes known at compile time) or in a vector otherwise.
+    using result_type = std::conditional_t<std::Same<search_scheme_t, search_scheme_dyn_type>,
+                                           std::vector<std::tuple<blocks_length_type, uint64_t>>,
+                                           std::array<std::tuple<blocks_length_type, uint64_t>,
+                                                      transformation_trait_or_t<std::tuple_size<search_scheme_t>,
+                                                                                std::false_type>::value>>;
+
+    bool constexpr is_dyn_scheme = std::Same<result_type, std::vector<std::tuple<blocks_length_type, uint64_t>>>;
```

Comment:
why not `std::Same<search_scheme_t, search_scheme_dyn_type>` and before defining the result_type so that you can use it there, too? Seems more logical...